### ISSUE TITLE
Nested hyperparameters cannot be updated

### ISSUE THREAD

**Issue by dpaetzel** [IQVIA-ML/TreeParzen.jl#104]
**Describe the bug**

I want to optimize the hyperparameters of a simple pipeline. According to the [MLJ docs](https://alan-turing-institute.github.io/MLJ.jl/dev/tuning_models/#Tuning-multiple-nested-hyperparameters), this should be possible by accessing these hyperparemeters using `:(dot.syntax.like.this)` (which works, e.g., for `Grid` tuning). However, this does not work with `MLJTreeParzenTuning`.

**To Reproduce**

An MWE is

```
using TreeParzen
using MLJ

standardizer = Standardizer()
DT = @load DecisionTreeRegressor pkg = DecisionTree
model = DT()
pipe = Pipeline(standardizer = standardizer, model = model)

X, y = make_regression(100, 5)

function usingexpr()
    space = Dict(:(model.max_depth) => HP.Choice(:(model.max_depth), collect(1:10)))
    pipe_tuned = TunedModel(
        model = pipe,
        tuning = MLJTreeParzenTuning(),
        range = space,
        measure = mae,
    )
    mach_tuned = machine(pipe_tuned, X, y)
    fit!(mach_tuned)
end

function usingsymbol()
    space = Dict(
        Symbol("model.max_depth") =>
            HP.Choice(Symbol("model.max_depth"), collect(1:10)),
    )
    pipe_tuned = TunedModel(
        model = pipe,
        tuning = MLJTreeParzenTuning(),
        range = space,
        measure = mae,
    )
    mach_tuned = machine(pipe_tuned, X, y)
    fit!(mach_tuned)
end

function usinggrid()
    space = range(pipe, :(model.max_depth); lower = 1, upper = 10)
    pipe_tuned = TunedModel(model = pipe, tuning = Grid(), range = space, measure = mae)
    mach_tuned = machine(pipe_tuned, X, y)
    fit!(mach_tuned)
end

# usingexpr()
# ERROR: LoadError: MethodError: no method matching TreeParzen.HP.Choice(::Expr, ::Vector{Int64})
#
# Closest candidates are:
#   TreeParzen.HP.Choice(::Symbol, ::Vector)
#    @ TreeParzen ~/.julia/packages/TreeParzen/gdoI0/src/HP.jl:116


# usingsymbol()
# ERROR: LoadError: Invalid hyperparameter: model.max_depth
# Stacktrace:
#   [1] error(s::String)
#     @ Base ./error.jl:35
#   [2] update_param!(model::MLJBase.DeterministicPipeline{NamedTuple{(:standardizer, :model), Tuple{Unsupervised, Deterministic}}, MLJModelInterface.predict}, param::Symbol, val::Int64)


usinggrid()
# Works.
```

**Expected behavior**

I expected to be able to tune nested hyperparameters just like when using the `Grid` strategy.

**Environment (please complete the following information):**
- OS: Linux
- Version: NixOS at `github:dpaetzel/nixpkgs/dpaetzel/nixos-config` .
- Julia Version: 1.9.3
- TreeParzen.jl Version: 0.3.3

**Additional context**

Thank you for making and maintaining this library! :slightly_smiling_face: 

It looks like the problem is caused by the [`update_param!` function in `MLJTreeParzen.jl`](https://github.com/IQVIA-ML/TreeParzen.jl/blob/master/src/MLJTreeParzen.jl#L221) not respecting nested hyperparameters.

**Comment by dpaetzel** [comment]
I'm currently looking into this. I think one could (and should?) reuse `MLJBase.recursive_setproperty!` but then the `Choice` (and other `HP`) constructor still would have to accept an `::Expr` which it does not righgt now. Would you be interested in a PR?

**Comment by kainkad** [comment]
Hi @dpaetzel thank you for raising this issue. I've checked your examples and the issue is not happening because of `update_params!` not supporting nested hyperparams but as you correctly observed the space with hyperparams to be optimised should be a dictionary where the keys (params) are Symbols and not Expressions or Symbols of those expressions.  Given your example, the model param should be `:max_depth` rather than `model.max_depth` or `Symbol("model.max_depth")`.
Based on your MWE, you created the pipeline object containing both the standardizer and the model and you passed that pipeline object to the `model` entry in the TunedModel() wrapper. This means that accessing the model params would then indeed require to be pipe.model.param but if instead you provided `model=pipe.model` then your space can be defined as below:

```
function usingexpr()
           space = Dict(
               :max_depth => HP.Choice(:max_depth, collect(1:10))
           )
           pipe_tuned = TunedModel(
               model = pipe.model,
               tuning = MLJTreeParzenTuning(),
               ranges = space,
               measure = mae,
           )
           mach_tuned = machine(pipe_tuned, X, y)
           fit!(mach_tuned)
       end
```
Would the above work for you?

If you’re interested in PR, then may be parsing the expression and extracting the hyperparm name to be passed to TreeParzen could be an option but this could be tricky if there were further changes to the MLJ's dot syntax.

**Comment by dpaetzel** [comment]
Thank you for the response! But doesn't `pipe.model` result in the tuning not having access to other parts of the pipeline? Example: I want to tune _both_ preprocessing and the actual model (e.g. if preprocessing is a feature selector with some tunable hyperparameters or similar).

**Comment by julien-kap** [comment]
I face the same issue: not being able to pass a composed pipeline with nested hyperparams. I tried to use the recursive getproperty and setproperty from MLJ into TreeParzen, but did not make it correctly.

**Comment by juliengooris** [comment]
I face the same issue: not being able to pass a composed pipeline with nested hyperparams. I tried to use the recursive getproperty and setproperty from MLJ into TreeParzen, but did not make it correctly.

**Comment by kainkad** [comment]
@dpaetzel Yes, this is correct that passing `pipe.model` will only have access to the model part of the pipeline which I think is in line with the MLJ's TunedModel wrapper where the model entry takes model not the whole pipe object as shown here: https://juliaai.github.io/MLJ.jl/dev/tuning_models/#Tuning-multiple-nested-hyperparameters, except for that that the range should be in a form of space dictionary required by TreeParzen.

However, I understand that being able to pass the whole pipeline object ensures that the tunning process considers the entire workflow including preprocessing and tuning them together may contribute to the better overall performance.

After fixing the initial formatting issue raised in your initial query which I explained in this comment https://github.com/IQVIA-ML/TreeParzen.jl/issues/104#issuecomment-2152013993, the actual problem now is that with the pipe object passed to the model entry of the TunedModel contains non tuneable entries in `pipe.standardizer`: 

```
standardizer = Standardizer(
        features = Symbol[],
        ignore = false,
        ordered_factor = false,
        count = false), 
```
so there's an error returned because TreeParzen can’t optimise that standardizer.

With your grid example, the space is defined using the `range()` function `space = range(pipe, :(model.max_depth); lower = 1, upper = 10)`, which takes the whole pipeline but only selects the tuneable param from the model part of that pipeline `model.max_depth`. Correct me if I'm wrong but the standardizer part is being ignored and the tuning only considers that model.max_depth param so technically it is the same as passing the model not the whole pipeline, or is the standardizer still being considered?
Would you may be have another example where the pipeline contains more than one tuneable entry from different parts of the pipeline which then can be passed to space being optimised? And how would you then define your space for that? At the moment with the standardizer, it would have been theoretically something like that I suppose (except for that the second range for standardizer.features won’t work due to unsupported values type):
```
space = range(pipe, :(model.max_depth); lower = 1, upper = 10) ∪
               range(pipe, :(standardizer.features); values = [Symbol[], [:feature1, :feature2]])
```
or defining separate ranges, `r1` and `r2` and passing them as a list in the `range=[r1, r2]`?

**Comment by dpaetzel** [comment]
Sorry, that I did not respond any more. I've since moved away from TPE and right now sadly do not have the resources to provide much more to this.

If I recall correctly, then in all my examples it was always a `Standardizer |> Regressor` pipeline (and tuning the standardizer hyperparameters obviously seldomly makes sense).

>  Correct me if I'm wrong but the standardizer part is being ignored and the tuning only considers that model.max_depth param so technically it is the same as passing the model not the whole pipeline, or is the standardizer still being considered?

Upon revisiting this: You may be right. The question is essentially whether `TunedModel(Standardizer |> Regressor) ≈ Standardizer |> TunedModel(Regressor)` and I suspect that since the standardizer parameters are not transformed it would probably make sense that this holds and the pipeline should simply be defined differently?

If it doesn't bother you that the semantics are different from `Grid()` then feel free to close this for now. 🙂

**Comment by kainkad** [comment]
Thank you @dpaetzel for looking into this despite of moving away from TPE. Yes, I think for now it's probably fine to close this issue.
